Object aware ResourceIdentifier

d-chambers opened the pull request:
### What does this PR do?

This pull request changes the default behavior of the ResourceIdentifier class slightly to allow for multiple objects to share the same resource id and still be accessible via the get_referred_object method. However, if no object was passed to either the __init__ method or the set_referred_object method, or if the original object goes out of scope, the behavior is the same as before in that the last defined object with the resource_id is returned.

### Why was it initiated?  Any relevant Issues?

Let me demonstrate with some code:

```python
import obspy
cat1 = obspy.read_events()
# The resource_id attached to the first event should point to the first event
cat1[0] is cat1[0].resource_id.get_referred_object()  # True
# but what happens if we make a copy? 
cat2 = cat1.copy()
# or perform another read operation?
cat3 = obspy.read_events()
# the resource_id on the new catalogs dont point to their attached objects
cat1[0] is cat1[0].resource_id.get_referred_object()  # True
cat2[0] is cat2[0].resource_id.get_referred_object()  # False
cat3[0] is cat3[0].resource_id.get_referred_object()  # False
# now what happens if the original catalog goes out of scope?
del cat1
cat2[0].resource_id.get_referred_object()  # None
cat3[0].resource_id.get_referred_object()  # None
# the resource_ids on the remaining objects are no longer usable
```

I have run into problems several times by using the get_referred_object method of the ResourceIdentifier class, both in testing and non-testing environments. For example, when trying to reference a pick from a phase often times the get_referred_object method returns none (even though the pick still exists in the current catalog), or harder to track yet, returns the pick from a copy of the catalog object. 

One solution, then, is to never copy an object with a resource_id or never read it in twice. This can, however, be impractical in some work flows, especially test environments that perform modifications to the object. The proposed modifications remove the surprising behavior and the code now executes as one (or at least me) might expect :

```python
import obspy

cat1 = obspy.read_events()
# The resource_id attached to the first event should point to the first event
cat1[0] is cat1[0].resource_id.get_referred_object()  # True
cat2 = cat1.copy()
cat3 = obspy.read_events()
# the resource_id on the new catalogs now point to their attached objects
cat1[0] is cat1[0].resource_id.get_referred_object()  # True
cat2[0] is cat2[0].resource_id.get_referred_object()  # True
cat3[0] is cat3[0].resource_id.get_referred_object()  # True
# now what happens if the original catalog goes out of scope? Nothing really...
del cat1
cat2[0].resource_id.get_referred_object()  # Event (the one you would expect)
cat3[0].resource_id.get_referred_object()  # Event
# now if we create a resource_id with the same id, the last defined object
# with the same resource id (that is still in scope) is returned
new_id = cat2[0].resource_id.id
rid = obspy.core.event.ResourceIdentifier(new_id)
rid.get_referred_object() is cat3[0]  # True
# if cat3 went out of scope then the object from cat2 would be returned
del cat3
rid.get_referred_object() is cat2[0]  # True
# if all objects with the same resource_id get gc’ed then None is returned
del cat2
rid.get_referred_object() is None  # True
```


### PR Checklist
- [x] This PR is not directly related to an existing issue (which has no PR yet).
- [x] All tests still pass.  (one fails but it failed on fresh fork)
- [x] Any new features or fixed regressions are be covered via new tests.
- [x] Any new or changed features have are fully documented.
- [x] Significant changes have been added to `CHANGELOG.txt` .
- [x] First time contributors have added your name to `CONTRIBUTORS.txt` .

krischer commented:
In general I quite like this but I'll have to review it in a bit more detail later - so thanks a lot! 

Also see this related PR which is likely obsolete due to your new PR as your solution is better: https://github.com/obspy/obspy/pull/1629

krischer commented on obspy/core/tests/test_event.py:
```diff
@@ -681,30 +717,22 @@ def test_same_resource_id_different_referred_object(self):
-        self.assertEqual(object_b is res_a.get_referred_object(), True)
-        self.assertEqual(object_b is res_b.get_referred_object(), True)
+        # This will no longer raise a warning
+        res_b = ResourceIdentifier(id=id, referred_object=object_b)
```
I'm fine with the behavior of the code but I think it should really still raise a warning if two non-equal objects share the same resource identifiers. The whole purpose of resource ids is to point to unique or at least equal objects. If that is not the case some logic is wrong and the users should be notified.

krischer commented on obspy/core/tests/test_event.py:
```diff
@@ -754,11 +782,47 @@ def test_adding_a_referred_object_after_creation(self):
+        warnings.simplefilter('default')
+        with warnings.catch_warnings(record=True) as w:
+            rid1.get_referred_object()
+            assert len(w)
```
Can you please add some fuzzy test of the warning message? Just to make sure that the correct warning is actually caught here.

krischer commented on obspy/core/tests/test_event.py:
```diff
@@ -773,12 +837,11 @@ def test_resources_in_global_dict_get_garbage_collected(self):
         self.assertEqual(len(list(rdict.keys())), 2)
-        # Deleting the objects should also remove the from the dictionary.
+        # Deleting the objects will no longer remove them from the dict, but
+        # the weak reference still behave as expected
```
Can you please change this comment? It will be a bit confusing without having the history of the file at hand.

d-chambers commented:
@krischer thanks for looking at this, happy to make the changes. 

Also, It occurred to me that this may not actually solve the problem of an arrival returning the wrong (non-event scoped) pick because the object_id attribute of the arrival.pick_id may never get set. One (ugly) solution might be to add a recursive function call on the Event.__init__ and Event.__deepcopy__ to find any ResourceIdentifier instances (not named resource_id) and set the object_id to that of the referred_object. Any thoughts? I should have time to work on this more in a few days.

krischer commented:
> Also, It occurred to me that this may not actually solve the problem of an arrival returning the wrong (non-event scoped) pick because the object_id attribute of the arrival.pick_id may never get set. One (ugly) solution might be to add a recursive function call on the Event.init and Event.deepcopy to find any ResourceIdentifier instances (not named resource_id) and set the object_id to that of the referred_object. Any thoughts? I should have time to work on this more in a few days.

Hmm...true. On the other hand `get_referred_object()` would still work but there is a chance that it return a pick from a different object which, by definition, should be equal to the event scoped pick. So yea - would be nice if this could be solved but no very big deal if not.

I'm not sure how to solve this to be honest - the ugly and tedious recursive solution will probably work for the deep copy but not for the normal .init case. Many people (and I'm pretty sure ObsPy internally also does this in lots of places) first create the event objects and only then fill it. 

We could create a `_internally_bind_resource_ids(event)` method that does this for a finished event and then automatically call this in `read_events()` so it will work in 99.9% of all cases.

krischer commented:
> Also, It occurred to me that this may not actually solve the problem of an arrival returning the wrong (non-event scoped) pick because the object_id attribute of the arrival.pick_id may never get set. One (ugly) solution might be to add a recursive function call on the Event.init and Event.deepcopy to find any ResourceIdentifier instances (not named resource_id) and set the object_id to that of the referred_object. Any thoughts? I should have time to work on this more in a few days.

Hmm...true. On the other hand `get_referred_object()` would still work but there is a chance that it return a pick from a different object which, by definition, should be equal to the event scoped pick. So yea - would be nice if this could be solved but no very big deal if not.

I'm not sure how to solve this to be honest - the ugly and tedious recursive solution will probably work for the deep copy but not for the normal .init case. Many people (and I'm pretty sure ObsPy internally also does this in lots of places) first create the event objects and only then fill it. 

We could create a `_internally_bind_resource_ids(event)` method that does this for a finished event and then automatically call this in `read_events()` so it will work in 99.9% of all cases.

d-chambers commented:
Ok, I think I addressed all of the issues but I am happy to make additional changes if needed.

I ended up adding another class dictionary to ResourceIdentifer that keeps track of all the ResourceIdentifer instances that have not been bound (ie had their object_id attribute set). I then added a ResourceIdentifier class method that iterates the dictionary and binds all of the unbound identifiers. I modified event level copy and the read_events function to call this method to get the "event scoped" behavior. Users wont have to know about this detail unless they manually create a catalog, as you described, in which case when finished they can simply do:
`resource_id.bind_resource_ids()`.
on any ResourceIdentifier instance.

krischer commented on obspy/core/event/base.py:
```diff
@@ -611,43 +667,73 @@ def get_referred_object(self):
         Will return None if no object could be found.
         """
```
I might be overthinking or over-complicating things here, but some of logic breaks down in a threaded context and there are multiple places where race conditions could happen.

For the resource ids that might actually happen fairly often as some users download in threads (I know I do). It might be worthwhile to add a class level `Threading.RLock` and wrap some methods in

```python
def get_referred_object(...):
    with self.__class__.__lock:
        ...
```

or maybe even a decorator? I think wrapping the `__init__()`, `get_referred_object()`, and `set_referred_object()` methods would take care of most race conditions.

d-chambers replied:
Ok, good point here but I am not sure how to do this correctly. In order to make this thread-safe(ish) I think the lock would need to be on read_events, and Catalog.copy as well wouldn't it? 

For example, if the following 3 things don't happen for each event before any occur for another event object with the same resource_ids, the "event scoped" behavior would be ruined:
1. Each non-resource_id object is created
2. The resource_id of each of the newly created objects calls set_referred_object 
3.  The _object_id attr of each any resource_ids not named "resource_id" is cleared
4. bind_resource_ids gets called 

If I add a lock on read_events and Catalog.copy wouldn't it drastically reduce the benefit you get from multi-threading for solving the small corner case of duplicate resource_ids? If we do create a decorator for thread locking, where should such a decorator live, in obspy.util.decorator? 

I dont understand how the race condition is created with the ResourceIdentifier methods you mentioned, although there probably is one I am just not seeing, as this is getting rather complicated. Thoughts?

krischer replied:
> Ok, good point here but I am not sure how to do this correctly. In order to make this thread-safe(ish) I think the lock would need to be on read_events, and Catalog.copy as well wouldn't it?

Yea - true thread safety with all references working for reading events would require that I think.

> If I add a lock on read_events and Catalog.copy wouldn't it drastically reduce the benefit you get from multi-threading for solving the small corner case of duplicate resource_ids? 

The performance will probably not deteriorate a lot (if at all) as the locking would just be around the actual event parsing routines. The downloading would still be threaded and this is likely the slow part. With CPython's GIL even most threaded things don't truly run in parallel (I/O or downloading is an exception).

> If we do create a decorator for thread locking, where should such a decorator live, in obspy.util.decorator?

Yes. I'm not sure if the effort can be justified though as we are currently just talking about the `obspy.core.event.catalog._read()` and the two copy methods - a `with` statement in these would suffice.

> I dont understand how the race condition is created with the ResourceIdentifier methods you mentioned, although there probably is one I am just not seeing, as this is getting rather complicated.

Locking the higher level function you proposed is much cleaner in any case. I'm too lazy now to try to re-follow my own thoughts about locking these three methods...I think it was related to the fact that all the class level dictionaries are not thread local and with all the required shuffling of object ids and references it might result in some non-deterministic race conditions. But as you said: its getting fairly complicated and I might be mistaken.

All in all this deals with a rare corner case but then its only 10 lines or so to do it properly so why not I guess?

krischer commented on obspy/core/event/base.py:
```diff
@@ -710,6 +796,18 @@ def copy(self):
         return deepcopy(self)
 
     @property
+    def object_id(self):
```
Can you add a leading underscore and designate it as private? I see no use to have this as a public part of the API and it frees us to make chances to it in the future if needed or desired.

krischer commented:
Thanks a lot for all the work!

I added another commit that makes sure the lock is created only once per function. The previous decorator created the lock at each function call which would result in no actual locking at all as each thread would just create its own lock. I think it now is correct but double checking does not hurt.

Otherwise this look IMHO pretty ready to merge.

+DOCS

d-chambers commented:
Great, thanks for your help and reviews.

I just realized the rlock should be on read_events not on the init method on the AbstractEventType, very small change.

krischer commented:
Great. Looks really good now. I had to rebase to get rid of a merge conflict. Thanks once again!

megies commented:
Reimplemented in #1702
